http-mclient/src/Network/HTTP/MClient.hs

```diff
@@ -138,16 +140,57 @@ send' transactionDecoder decoder request = do
   sendRequest sendRequest' transactionDecoder decoder request
 
 sendRequest
-  :: SendRequest
+  :: forall env a . Env env
+  => SendRequest
   -> TransactionDecoder a
   -> ResponseDecoder a
   -> HTTP.Request
   -> MIO env (Result a)
 sendRequest sendRequest' transactionDecoder decoder request =
-  either transactionDecoder decoder
-    <$> Exception.tryJust selectException (liftIO $ sendRequest' request)
+  either transactionDecoder decoder <$> retriableSendRequest
   where
-    selectException :: HTTP.HttpException -> Maybe HTTP.HttpExceptionContent
-    selectException = \case
-      (HTTP.HttpExceptionRequest _request content) -> pure content
-      _other                                       -> empty
+    retriableSendRequest :: MIO env (Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString))
+    retriableSendRequest
+      = Retry.retrying
+        limitedBackoff
+        isRetriable
+      $ const (Exception.tryJust selectException . liftIO $ sendRequest' request)
+      where
+        limitedBackoff :: Retry.RetryPolicy
+        limitedBackoff = Retry.exponentialBackoff 50_000 <> Retry.limitRetries (convertImpure retryLimit)
+
+        isRetriable
+          :: Retry.RetryStatus
+          -> Either HTTP.HttpExceptionContent (HTTP.Response LBS.ByteString)
+          -> MIO env Bool
+        isRetriable Retry.RetryStatus{..} = either process (const (pure False))
+          where
+            process :: HTTP.HttpExceptionContent -> MIO env Bool
+            process error = if isRequestNotSent error then logError $> True else pure False
+              where
+                isRequestNotSent = \case
+                  HTTP.ConnectionTimeout            -> True
+                  HTTP.ConnectionClosed             -> True
```

Comment:
If the connection was unexpectedly closed do we have the guarantee that the request never made it to the server?

Reply:
No, we don't have a `guarantee`. What I can tell is that we seem to have closed connections only during reads from the servers. At that point, a request was already made. I will remove it in the next update.